### ISSUE TITLE
Please allow trailing text on "clang-format on/off" comments

### ISSUE THREAD

**Issue by alexolog** [llvm/llvm-project#60264]
Currently, any additional text after the "off" prevents the directive from being processed, so we have to resort to ugly workarounds in order to provide an explanation for turning the formatting off.

Real life example:
```cpp
/* clang-format off */ // version 14 messes up concepts formatting
template<typename T>
concept ProtoContainer = requires(T x) {
    { x.mutable_elements() };
};
// clang-format on
```

Desired outcome: the following should work
```cpp
// clang-format off - reasoning and explanation here
...
// clang-format on
```

Note that clang-tidy allows extra text on `NOLINT` comments
(see https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clang-tidy/NoLintDirectiveHandler.cpp)

**Comment by owenca** [comment]
https://reviews.llvm.org/D142804

**Comment by alexolog** [comment]
Here's my attempt at something flexible:
Disclaimer: this is my first time looking at the LLVM codebase, so don't expect production quality.

```
////////////////////////////////////////////////////////////////////////////////////////////////////

// Implementation detail, hide in a namespace and/or take out of the header
bool isClangFormatMarked(StringRef Comment, StringRef Mark) {
  // Quick heuristics: minimum length and starts with a slash (comment)
  // Shortest tag: "//clang-format on", 17 characters
  static constexpr StringLiteral clangFormatStr("clang-format ");
  if (Comment.size() < clangFormatStr.size() + 4 || Comment[0] != '/')
    return false;

  // check if it's a comment starting with "//" or "/*"
  bool CloseNeeded = false;
  if (Comment[1] == '*')
    CloseNeeded = true;
  else if (Comment[1] != '/')
    return false;

  // Remove the comment start and all following whitespace
  Comment = Comment.substr(2).ltrim();

  // Check for the actual command, a piece at a time
  if (!Comment.consume_front(clangFormatStr) || !Comment.consume_front(Mark))
    return false;

  // Are we there yet?
  if (!CloseNeeded && Comment.empty() ||
      CloseNeeded && Comment.starts_with("*/"))
    return true;

  // For a trailer, restrict the next character
  // (currently spaces and tabs, but can include a colon, etc.)
  static constexpr StringLiteral Separator(" \t");
  if (!Separator.contains(Comment[0]))
    return false;
  
  // Verify comment is properly terminated
  if (!CloseNeeded || Comment.contains("*/"))
    return true;

  return false; // Everything else
}

////////////////////////////////////////////////////////////////////////////////////////////////////

bool isClangFormatOn(StringRef Comment) {
  return isClangFormatMarked(Comment, "on");
}

bool isClangFormatOff(StringRef Comment) {
  return isClangFormatMarked(Comment, "off");
}

////////////////////////////////////////////////////////////////////////////////////////////////////

EXPECT_TRUE(isClangFormatOn("//clang-format on"));
EXPECT_TRUE(isClangFormatOn("// clang-format on"));
EXPECT_TRUE(isClangFormatOn("//clang-format on "));
EXPECT_TRUE(isClangFormatOn("//clang-format on and off"));
EXPECT_TRUE(isClangFormatOn("/*clang-format on*/"));
EXPECT_TRUE(isClangFormatOn("/* clang-format on*/"));
EXPECT_TRUE(isClangFormatOn("/*clang-format on */"));
EXPECT_TRUE(isClangFormatOn("/*clang-format on*/int i{0};"));

EXPECT_FALSE(isClangFormatOn("//clang-format  on"));
EXPECT_FALSE(isClangFormatOn("//clang-format o"));
EXPECT_FALSE(isClangFormatOn("//     clang-format o"));
EXPECT_FALSE(isClangFormatOn("//clang-format ontario"));
EXPECT_FALSE(isClangFormatOn("//clang-format off"));
EXPECT_FALSE(isClangFormatOn("/*clang-format onwards*/"));

////////////////////////////////////////////////////////////////////////////////////////////////////
```

**Comment by bersbersbers** [comment]
Stumbled across this by accidently today (literally a day after it landed!) on https://clang.llvm.org/docs/ClangFormatStyleOptions.html:

![image](https://user-images.githubusercontent.com/12128514/216538685-01c442f6-7d69-460d-8eef-dffacfca1089.png)

Seem there is an extra space in one of the double backticks (middle of bottom line).

**Comment by owenca** [comment]
Thanks! Fixed in 4f37de9c91eb.
